src/ray/gcs/redis_context.cc

```diff
@@ -595,16 +597,27 @@ Status RedisContext::Connect(const std::string &address,
 
   RAY_CHECK(!context_);
   RAY_CHECK(!redis_async_context_);
+
+  // Remember function arguments for reconnection
+  port_ = port;
+  username_ = username;
+  password_ = password;
+  enable_ssl_ = enable_ssl;
+
   // Fetch the ip address from the address. It might return multiple
   // addresses and only the first one will be used.
-  auto ip_addresses = ResolveDNS(address, port);
-  RAY_CHECK(!ip_addresses.empty())
-      << "Failed to resolve DNS for " << address << ":" << port;
+  if (ip_address_.empty()) {
+    RAY_CHECK(!address.empty());
+    auto ip_addresses = ResolveDNS(address, port);
```

Comment:
We need to `ResolveDNS` very time we connect, otherwise we don't find the new master if the old one dies

Reply:
I'm not sure I follow, we don't clear DNS cache anywhere, how would re-resolve change?

Reply:
fixed

Reply:
> I'm not sure I follow, we don't clear DNS cache anywhere, how would re-resolve change?

OS should clear the DNS cache based on TTL?

Reply:
TTL is not function invocation magnitude. see my comment above
https://github.com/ray-project/ray/pull/48781#discussion_r1874396312

Reply:
Let me try this way, usually HA redis instance is hosted under an API server (i.e. GCP memstore), the IP address and DNS here is used to resolve and connect to the API server, rather than the actual instance.

Just like you deploying a microservice with k8s service, you won't change the IP for your pod failure, right?
service discovery inside of k8s coredns IIRC

> OS should clear the DNS cache based on TTL?

The timeout is pretty long, and apart from your OS, network devices along the way has DNS cache as well, which you don't have control over. 
We used to move a k8s cluster from us-west1 region to us-central1 region for a service under configured A record, it took a whole day to route 50% of traffic.